### PR TITLE
[feature] #2073: Prefer ConstString over String

### LINKED ISSUE #2073: Replace immutable strings with `ConstString`
In some places, namely `data_model` and `crypto` library there are immutable structures which store `String`s that can never be modified. Examples include `PublicKey`, `PrivateKey`, `IpfsPath`, ...

Grep the codebase for `String`s and make use of the newly added `ConstString` instead of `std::String` where applicable to reduce the memory footprint. #1985

### PR DESCRIPTION
**Pull request by ales-tsurko**
### Description of the Change

Replaces `String` with `ConstString` for types stored in the blockchain. It also encapsulates `iroha_data_primitives` into a shared crate.

### Issue

Closes #2073

### CODE REVIEW

**Comment by mversic** [comment]
we don't need `ConstString` in tests. We also don't need it when returning errors. It mostly makes sense to use it for things that are stored in memory and never modified. An example of this is `PublicKey::digest`

**Comment by ales-tsurko** [comment]
> we don't need `ConstString` in tests. We also don't need it when returning errors. It mostly makes sense to use it for things that are stored in memory and never modified. An example of this is `PublicKey::digest` in `iroha_crypto` or `IpfsPath` in `data_model`

OK, I'll remove it in tests and errors where possible, but somewhere it's updated because of dependencies.

>  It mostly makes sense to use it for things that are stored in memory and never modified. 

This part of the task is not very clear to me. For example, we initialize an error with string, this string is stored in memory and it's never modified afterwards. Actually, there's a very rare case when we need to modify `String`, so these requirements can be applied to a lot of types, including errors and those from the tests.

**Comment by mversic** [comment]
> This part of the task is not very clear to me. For example, we initialize an error with string, this string is stored in memory and it's never modified afterwards. Actually, there's a very rare case when we need to modify String, so these requirements can be applied to a lot of types, including errors and those from the tests.

what you say is true but the benefit is marginal. We need `ConstString` for types that are stored on the blockchain

**Review by mversic — APPROVED** [review]
Looks good
I see you dropped implementation on `Peer::address`. I support this revert.

File: Cargo.toml [hunk 1]
```diff
@@ -12,11 +12,8 @@ members = [
   "core",
   "core/test_network",
   "crypto",
-  "tools/kura_inspector",
-  "tools/parity_scale_decoder",
-  "tools/kagami",
   "data_model",
-  "data_model/primitives",
+  "data_primitives",
```
[the remark this answers is not in the document]
Comment (ales-tsurko) [reply Cargo.toml:16]: I'd rename it into `primitives` as well. Just kept existing name. Should I rename it?

[the remark this answers is not in the document]
Comment (mversic) [reply Cargo.toml:16]: > Will revert `PeerId::address`, ok

I don't think there is anything to change. All is good

[the remark this answers is not in the document]
Comment (ales-tsurko) [reply Cargo.toml:16]: > > Will revert `PeerId::address`, ok
> 
> I don't think there is anything to change now. All is good

Ah, sorry for misunderstanding.

What about renaming `data_primitives` into `primitives`?

[the remark this answers is not in the document]
Comment (mversic) [reply Cargo.toml:16]: > What about renaming data_primitives into primitives?

Now I see `data_model/primitives` was already a separate crate. I think you could have referenced it directly from `iroha_crypto` without moving it outside of `data_model`?

File: data_model/tests/ui_fail/fix_num_private.rs [hunk 2]
```diff
@@ -1,4 +1,4 @@
-use iroha_data_model::fixed::{Fixed, FixNum};
+use iroha_data_primitives::fixed::{Fixed, FixNum};
```
[the remark this answers is not in the document]
Comment (mversic) [reply data_model/tests/ui_fail/fix_num_private.rs:1]: Oh, yes. Good call

[the remark this answers is not in the document]
Comment (ales-tsurko) [reply data_model/tests/ui_fail/fix_num_private.rs:1]: Oh, thanks

**Review by mversic — APPROVED** [review]

**Review by mversic — APPROVED** [review]

**Review by Erigara — APPROVED** [review]

**Review by mversic — APPROVED** [review]

**Review by Erigara — APPROVED** [review]

**Pull request merged**